File: backend/cpfcnpj-crud-api/query/model/query.go
package model

import (
	"encoding/json"
)

const (
	SortFieldOrderASC  = "asc"
	SortFieldOrderDESC = "desc"
)

type Query struct {
	Page           int64  `query:"page"`
	ResultsPerPage int64  `query:"resultsperpage"`
	SortsJson      string `query:"sorts"`
	FiltersJson    string `query:"filters"`
	Sorts          []*SortField
	Filters        []*FilterField
}

type SortField struct {
	Name  string `json:"name"`
	Order string `json:"order"`
}

type FilterField struct {
	Name  string      `json:"name"`
	Value interface{} `json:"value"`
}

/*
	Url params for sorts and filters come as json values inside string field "SortsJson" and "FiltersJson".
	With this function, these string will be a array of objects.
*/
func (q *Query) ParseJsonParams() {
	if len(q.FiltersJson) > 0 {
		var filters []*FilterField
		errors := json.Unmarshal([]byte(q.FiltersJson), &filters)

		if errors == nil {
			q.Filters = filters
		}
	}

	if len(q.SortsJson) > 0 {
		var sorts []*SortField
		errors := json.Unmarshal([]byte(q.SortsJson), &sorts)

		if errors == nil {
			q.Sorts = sorts
		}
	}
}

File: frontend/cpfcnpj-crud-app/src/model/Query.ts
export default interface Query {
  page: number;
  resultsPerPage: number;
  sorts: SortField[] | null;
  filters: FilterField[] | null;
  toParam(): any;
}

export interface SortField {
  name: string;
  order: string;
}

export interface FilterField {
  name: string;
  value: any;
}

export const emptyQuery: Query = {
  page: 0,
  resultsPerPage: 10,
  sorts: null,
  filters: null,
  toParam() {
    return {
      page: this.page,
      resultsPerPage: this.resultsPerPage,
      sorts: this.sorts != null ? JSON.stringify(this.sorts) : null,
      filters: this.filters != null ? JSON.stringify(this.filters) : null,
    };
  },
};
